src/recorder.ts

```diff
@@ -0,0 +1,130 @@
+import { EventEmitter2 } from 'eventemitter2';
+import { IExtendParams, IMatcher } from './matcher';
+import { IObserver } from './observers';
+import { MatcherKey, Step, StepEvent } from './types';
+import { BasicMetaQuerier, IMetaQuerier } from './util/metaquerier';
+
+type StepEventHandler = (step: Step) => void;
+
+export type RecorderOptions = {
+  matcher: IMatcher;
+  onEmit: StepEventHandler;
+  metaQuerier?: IMetaQuerier;
+};
+
+export class Recorder {
+  private observersList: IObserver[] = [];
+  private emitterMap: WeakMap<IObserver, EventEmitter2> = new WeakMap();
+
+  private matcher: IMatcher;
+
+  private onEmit: StepEventHandler;
+  private metaQuerier: IMetaQuerier;
+
+  private state: 'active' | 'inactive' | 'suspend' = 'inactive';
+
+  constructor(options: RecorderOptions) {
+    const { matcher, onEmit, metaQuerier } = options;
+    this.matcher = matcher;
+    this.matcher.emitter = new EventEmitter2();
+    this.onEmit = onEmit;
+    this.matcher.emitter.on(
+      MatcherKey.EMIT,
+      (
+        action: Step['action'] | 'UNKNOWN',
+        events: StepEvent[],
+        target: HTMLElement | null,
+      ) => {
+        this.emitCurrentStep(action, events, target);
+      },
+    );
+    this.metaQuerier = metaQuerier || new BasicMetaQuerier();
+  }
+
+  public start(): void {
+    this.state = 'active';
+    this.observersList.forEach((obs) => {
+      obs.start();
+    });
+    this.matcher.start();
+  }
+
+  public suspend(): void {
+    this.state = 'suspend';
+    this.observersList.forEach((obs) => {
+      obs.suspend();
+    });
+    this.matcher.suspend();
+  }
+
+  public stop(): void {
+    this.state = 'inactive';
+    this.observersList.forEach((obs) => {
+      obs.stop();
+    });
+    this.matcher.stop();
+  }
+
+  public extendAction<params extends IExtendParams>(action: params): IObserver {
+    const { observer } = action;
+    if (this.state === 'active') {
+      console.warn(
+        'cannot extend recorder when active, please suspend or stop recorder first',
+      );
+      return observer;
+    }
+    if (this.emitterMap.has(observer)) {
```

Comment:
we can have only one listener on an event ?

Reply:
It means one observer should only have a single group processing method.
When extended, they are immutable.
If you want to modify the action, you should firstly suspend the the recorder and remove this action and re-extend the observer.